### PR TITLE
Proposal for _unlockNextOperation parameter

### PR DESCRIPTION
**Pull request by allamand**
This is a proposal to implement a new ForceNewOperation parameter in order to allow to recover from an endless pending cluster state:

- if there is not enough ressource to shedule the requested pods
- if there are some selector labels which are not good

### CODE REVIEW

File: documentation/troubleshooting.md [hunk 1]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
 $ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=cluster-admin --user=myname@example.org
 Clusterrolebinding "myname-cluster-admin-binding" created
 ```
+
+### Pod and volumes can be scheduled in different zones using default provisioned
+
+The default provisioner in GKE don't have the `volumeBindingMode: "WaitForFirstConsumer"` option that can result in bad
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `don't` -> `does not`
`in bad` -> `in a bad`

File: documentation/troubleshooting.md [hunk 2]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
...
+scheduling behaviour.
+I uses files to create storageclass:
+- samples/gke-storage-standard-wait.yaml
+- samples/gke-storage-ssd-wait.yaml (if you have ssd disks)
+
+
+## Operator can't perform the Action
+
+If you ask to ScaleUp or add a new DC, or ask for more ressources, CassKop will ask Kubernetes to schedule as you
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: let's say `scale up` instead of using the function that we call internally

File: documentation/troubleshooting.md [hunk 3]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
...
+If we looked at the pod status we will see this message :
+
+```
+Warning  FailedScheduling   51s (x17 over 14m)    default-scheduler   0/6 nodes are available: 4 Insufficient cpu, 4 node(s) didn't match node selector.
+```
+
+Kubernetes can't find any Pod with sufficiant cpu and matching kubernetes nodes labels we askde in the topology section.
+
+To fix this, we can either:
+- add more kubernetes nodes that will satisfied our requirements.
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: We can also potentially reduce memory/cpu limits

File: documentation/troubleshooting.md [hunk 4]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
...
+
+the cassandra-demo-dc1-rack1-1 pod is Pending and can't be scheduled.
+
+If we looked at the pod status we will see this message :
+
+```
+Warning  FailedScheduling   51s (x17 over 14m)    default-scheduler   0/6 nodes are available: 4 Insufficient cpu, 4 node(s) didn't match node selector.
+```
+
+Kubernetes can't find any Pod with sufficiant cpu and matching kubernetes nodes labels we askde in the topology section.
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `askde` -> `asked`

File: documentation/troubleshooting.md [hunk 5]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
...
+INFO[3354] Decommission done -> we delete PVC            cluster=cassandra-demo pvc=data-cassandra-demo-dc1-rack1-1 rack=dc1-rack1
+INFO[3354] PVC deleted                                   cluster=cassandra-demo pvc=data-cassandra-demo-dc1-rack1-1 rack=dc1-rack1
+DEBU[3354] Waiting Rack to be running before continuing, we break ReconcileRack Without Updating Statefulset  cluster=cassandra-demo dc-rack=dc1-rack1 err="<nil>"
+INFO[3354] ScaleDown is Done                             cluster=cassandra-demo rack=dc1-rack1
+```
+
+
+### Example: can't add new rack in new DC
+
+In this example, I ask to ad a new dc dc2
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `add`, `dc called dc2`

File: pkg/controller/cassandracluster/cassandra_util.go [hunk 6]
```diff
@@ -34,6 +34,10 @@ func (rcc *ReconcileCassandraCluster) thereIsPodDisruption() bool {
 	return rcc.storedPdb.Status.PodDisruptionsAllowed == 0
 }
 
+func (rcc *ReconcileCassandraCluster) weAllowMoreThan1PodDisruption() bool {
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/cassandra_util.go]: Let's rename that one to `allowPodDisruption`

File: pkg/controller/cassandracluster/cassandra_util.go [hunk 7]
```diff
@@ -34,6 +34,10 @@ func (rcc *ReconcileCassandraCluster) thereIsPodDisruption() bool {
...
+	return rcc.storedPdb.Spec.MaxUnavailable.IntVal > 1
+}
+
 func (rcc *ReconcileCassandraCluster) thereIsOnly1PodDisruption() bool {
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/cassandra_util.go]: `hasOneDisruptedPod` ?

File: pkg/controller/cassandracluster/pod.go [hunk 8]
```diff
@@ -38,6 +38,20 @@ const (
 
 var reEndingNumber = regexp.MustCompile("[0-9]+$")
 
+// allPodContainerReady return true if all container in the Pod are ready
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod.go]: `returns`

File: pkg/controller/cassandracluster/pod.go [hunk 9]
```diff
@@ -38,6 +38,20 @@ const (
...
+func AllPodContainerReady(pod *v1.Pod) bool {
+	if pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0{
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod.go]: I think `pod.Status.ContainerStatuses` is at least set to an empty list by default ? if I'm right, we don't need to test if it's nil

File: pkg/controller/cassandracluster/pod.go [hunk 10]
```diff
@@ -38,6 +38,20 @@ const (
 
 var reEndingNumber = regexp.MustCompile("[0-9]+$")
 
+// allPodContainerReady return true if all container in the Pod are ready
+func AllPodContainerReady(pod *v1.Pod) bool {
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod.go]: `PodContainersReady` ?

File: pkg/controller/cassandracluster/pod.go [hunk 11]
```diff
@@ -111,6 +122,21 @@ func (rcc *ReconcileCassandraCluster) UpdatePodLabel(pod *v1.Pod, label map[stri
 	return rcc.UpdatePod(podToUpdate)
 }
 
+func (rcc *ReconcileCassandraCluster) isAPodUnschedulable(namespace string, dcName, rackName string) bool {
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod.go]: `hasUnschedulablePod` ?

File: pkg/controller/cassandracluster/pod.go [hunk 12]
```diff
@@ -111,6 +122,21 @@ func (rcc *ReconcileCassandraCluster) UpdatePodLabel(pod *v1.Pod, label map[stri
...
+	podsList, err := rcc.ListPods(rcc.cc.Namespace, k8s.LabelsForCassandraDCRack(rcc.cc, dcName, rackName))
+	nb := len(podsList.Items)
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod.go]: `nb` is only used once. Let's inline the code instead of creating a variable

File: pkg/controller/cassandracluster/pod_test.go [hunk 13]
```diff
@@ -82,16 +83,48 @@ func TestGetLastOrFirstPod(t *testing.T) {
...
+	}
 	assertPodIsNotRunning(t, err)
 }
 
+
+func TestAllPodContainerReady(t *testing.T) {
+
+	mkPod := func(id int, containers ...bool) *v1.Pod {
+		containerStatuses := []v1.ContainerStatus{}
+	var i int = 0
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod_test.go]: let's use a better name for that variable and indent the code properly

File: pkg/controller/cassandracluster/pod_test.go [hunk 14]
```diff
@@ -82,16 +83,48 @@ func TestGetLastOrFirstPod(t *testing.T) {
...
+			containerStatuses = append(containerStatuses, containerStatus)
+		}
+
+		return &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("ns-nm-%02d", id)},
+			Status: v1.PodStatus{Phase: v1.PodRunning, ContainerStatuses:containerStatuses}}
+	}
+
+	assert.Equal(t, true, AllPodContainerReady(mkPod(1, true)))
+	//No container can't be ready
+	assert.Equal(t, false, AllPodContainerReady(mkPod(1, )))
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod_test.go]: empty parameter here ?

File: documentation/troubleshooting.md [hunk 15]
```diff
@@ -28,3 +40,189 @@ Account: [myname@example.org]
...
+let's remove the rack3 of dc2.
+
+the operator will log :
+
+```
+WARN[0347] We asked to remove Rack dc2-rack3 with unschedulable pod  cluster=cassandra-demo
+INFO[0347] [cassandra-demo]: Delete PVC[data-cassandra-demo-dc2-rack3-0] OK 
+```
+
+The rack3 (and it's statefulset) has been removed, and the associated (empty) pvc deleted
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `it's` -> `its`

File: pkg/controller/cassandracluster/cassandra_status.go [hunk 16]
```diff
@@ -76,21 +76,40 @@ func (rcc *ReconcileCassandraCluster) getNextCassandraClusterStatus(cc *api.Cass
 		return nil
 	}
 
-	//Do nothing in Initial phase
+	//If we set up ForceNewOperation in CRD we allow to see mode change even last operation didn't ended correctly
+	var needSpecificChange bool = false
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/cassandra_status.go]: just do `needSpecificChange := true`

File: pkg/controller/cassandracluster/cassandra_status.go [hunk 17]
```diff
@@ -405,13 +424,32 @@ func (rcc *ReconcileCassandraCluster) UpdateCassandraRackStatusPhase(cc *api.Cas
...
+			logrus.WithFields(logrus.Fields{"cluster": cc.Name,
+				"rack": dcRackName}).Warn("Aborting Initializing..., start ScaleDown")
+			status.CassandraRackStatus[dcRackName].Phase = api.ClusterPhasePending
+			now := metav1.Now()
+			lastAction.StartTime = &now
+			lastAction.Status = api.StatusToDo
+			lastAction.Name = api.ActionScaleDown
+			status.CassandraRackStatus[dcRackName].PodLastOperation.Status = api.StatusToDo
+			status.CassandraRackStatus[dcRackName].PodLastOperation.Name = api.OperationDecommission
+			status.CassandraRackStatus[dcRackName].PodLastOperation.StartTime = &now
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/cassandra_status.go]: I think we set all the following parameters in different places the same way. We should probably refactor that code

File: pkg/controller/cassandracluster/cassandra_util.go (hunk 6 above)
Comment (allamand) [reply pkg/controller/cassandracluster/cassandra_util.go]: go pout allowMoreThan1PodDisruption (cf le test)

File: pkg/controller/cassandracluster/pod.go (hunk 9 above)
Comment (allamand) [reply pkg/controller/cassandracluster/pod.go]: I had some panic while accessing the content of containerStatuses if it was not nil so the test seams necessary

File: pkg/controller/cassandracluster/pod_test.go (hunk 13 above)
Comment (allamand) [reply pkg/controller/cassandracluster/pod_test.go]: I took that name from your code :)

File: pkg/controller/cassandracluster/pod_test.go (hunk 14 above)
Comment (allamand) [reply pkg/controller/cassandracluster/pod_test.go]: that was pod with no container, i removed the ,

File: documentation/troubleshooting.md (hunk 5 above)
Comment (cscetbon) [reply documentation/troubleshooting.md]: @allamand ^

File: documentation/troubleshooting.md (hunk 3 above)
Comment (cscetbon) [reply documentation/troubleshooting.md]: @allamand ^

File: pkg/controller/cassandracluster/cassandra_status.go [hunk 18]
```diff
@@ -490,3 +471,20 @@ func (rcc *ReconcileCassandraCluster) UpdateCassandraRackStatusPhase(cc *api.Cas
 	}
 	return nil
 }
+
+
+func setDecommissionStatus(status *api.CassandraClusterStatus, dcRackName string){
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/cassandra_status.go]: Far better !

File: pkg/controller/cassandracluster/pod_test.go [hunk 19]
```diff
@@ -104,25 +104,26 @@ func TestAllPodContainerReady(t *testing.T) {
...
 
 		return &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("ns-nm-%02d", id)},
 			Status: v1.PodStatus{Phase: v1.PodRunning, ContainerStatuses:containerStatuses}}
 	}
 
-	assert.Equal(t, true, AllPodContainerReady(mkPod(1, true)))
+	assert.Equal(t, true, PodContainersReady(mkPod(1, true)))
 	//No container can't be ready
-	assert.Equal(t, false, AllPodContainerReady(mkPod(1, )))
+	assert.Equal(t, false, PodContainersReady(mkPod(1 )))
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/pod_test.go]: @allamand remove extra trailing space

**Comment by cscetbon** [comment]
@allamand what about the parameter name ? I don't really like that name. What about starting it with an `_` to make it a special parameter we should almost never use ?

File: documentation/troubleshooting.md [hunk 20]
```diff
@@ -28,3 +40,192 @@ Account: [myname@example.org]
 $ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=cluster-admin --user=myname@example.org
 Clusterrolebinding "myname-cluster-admin-binding" created
 ```
+
+### Pod and volumes can be scheduled in different zones using default provisioned
+
+The default provisioner in GKE does not have the `volumeBindingMode: "WaitForFirstConsumer"` option that can result in
+a bad
+scheduling behaviour.
+I uses files to create storageclass:
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `I use one of the following files to create a storage clas`

File: documentation/troubleshooting.md (hunk 3 above)
Comment (allamand) [reply documentation/troubleshooting.md]: ^^

**Comment by allamand** [comment]
@cscetbon it's not possible to preffix with _ because the var won't be accessible anymore. But we can find a better name ?

**Comment by cscetbon** [comment]
@allamand let's use the new name we both prefer and fix the current conflicts 👆

File: documentation/troubleshooting.md [hunk 21]
```diff
@@ -28,3 +40,193 @@ Account: [myname@example.org]
...
+
+#### Rollback ScaleUp operation
+
+In order to rollback the operation, we need to revert the change on the `nodesPerRacks` parameter.
+
+> This is not sufficient
+
+Because CassKop is actually performing another action on the cluster (ScaleUp) we can't scheduled a new operation to
+rollback since it has not finished.
+We introduced a new parameter in the CRD to allow such changes when all the pods can't be scheduled:
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: @allamand let's add a 🚩 to show it's not a regular parameter

File: pkg/controller/cassandracluster/node_operations_test.go [hunk 22]
```diff
@@ -484,7 +484,7 @@ func TestHostIDMap(t *testing.T) {
 			"value": {"10.244.3.20": "ac0b9f2b-1eb4-40ca-bc6e-68b37575f019"},
 			"timestamp": 1528850319,
 			"status": 200}`))
-	jolokiaClient, _ := NewJolokiaClient(host, JolokiaPort, nil,
+	jolokiaClient, _ := NewJolokiaClient(host, cassandraJolokia, nil,
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/node_operations_test.go]: @allamand Why `CassandraJolokia` ?? Jolokia is used to access JMX even if it's not used with Cassandra

Comment (allamand) [reply pkg/controller/cassandracluster/node_operations_test.go]: the vars was declared in 2 different places, I keep the name of the second.. I can rename it

Comment (cscetbon) [reply pkg/controller/cassandracluster/node_operations_test.go]: Yeah I would keep `JolokiaPort` instead. It'll also make things simpler for the pending PRs.

Comment (allamand) [reply pkg/controller/cassandracluster/node_operations_test.go]: ok I change it

**Comment by allamand** [comment]
@cscetbon Can we merge now ?

File: documentation/troubleshooting.md [hunk 23]
```diff
@@ -28,3 +40,196 @@ Account: [myname@example.org]
...
+- `Spec._unlockNextOperation: true`
+
+> _**:triangular_flag_on_post: Warning** This is not a regular parameter and it must be used with very good care!!._
+
+
+By adding this parameter in our cluster definition, CassKop will allow to trigger a new operation.
+
+> Once CassKop will scheduled the new operation, it will reset the `Spec._unlockNextOperation: false` to the default false
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: @allamand Once CassKop `has` scheduled the new operation, it will reset this parameter to the default `false` value

File: documentation/troubleshooting.md [hunk 24]
```diff
@@ -28,3 +40,196 @@ Account: [myname@example.org]
...
+        name: rack1
+      - labels:
+          failure-domain.beta.kubernetes.io/zone: europe-west1-c
+        name: rack2
+      - labels:
+          failure-domain.beta.kubernetes.io/zone: europe-west1-d
+        name: rack3
+```
+
+let's remove the rack3 of dc2.
```
Comment (cscetbon) [inline documentation/troubleshooting.md]: `let's remove rack3 from dc2`

File: pkg/controller/cassandracluster/reconcile.go [hunk 25]
```diff
@@ -180,8 +180,50 @@ func CheckNonAllowedRemoveDC(rcc *ReconcileCassandraCluster, cc *api.CassandraCl
...
+								cc.Spec.Topology = oldCRD.Spec.Topology
+								return true, api.ActionCorrectCRDConfig
+							}
+							rcc.DeletePVCs(cc, dcName, rackName)
+							return true, api.ActionDeleteRack
+						}
+					}
+
+			}
+			logrus.WithFields(logrus.Fields{"cluster": cc.Name}).Warningf("The Operator has refused the Topology changed. You can only remove an entire DC, not only a Rack: %v restored to %v", cc.Spec.Topology, oldCRD.Spec.Topology)
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/reconcile.go]: Such lines are far too long, same for below and above. Let's split them over 2 lines

File: pkg/controller/cassandracluster/reconcile.go [hunk 26]
```diff
@@ -180,8 +180,50 @@ func CheckNonAllowedRemoveDC(rcc *ReconcileCassandraCluster, cc *api.CassandraCl
...
+
+						listPod, err := rcc.ListPods(cc.Namespace, k8s.LabelsForCassandraDCRack(cc, dcName, rackName))
+						if err != nil {
+							logrus.WithFields(logrus.Fields{"cluster": cc.Name}).Warningf(
+								"The Operator has refused the Topology changed. We can't get associated pod in rack Rack: %v restored to %v", cc.Spec.Topology, oldCRD.Spec.Topology)
+							cc.Spec.Topology = oldCRD.Spec.Topology
+							return true, api.ActionCorrectCRDConfig
+						}
+						if len(listPod.Items) > 1{
+							logrus.WithFields(logrus.Fields{"cluster": cc.Name}).Warning("The Operator has refused the Topology changed. Too many pods in the Statefulset")
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/reconcile.go]: @allamand, When could that one happen again ?

Comment (cscetbon) [reply pkg/controller/cassandracluster/reconcile.go]: oh it can happen if the scale down failed or did not happen ?

Comment (allamand) [reply pkg/controller/cassandracluster/reconcile.go]: it happens in case we added a new rack that has not been created yet because of lack of ressources. In this case it is allowed to remove the rack, but only if there is no pod at all really created in this rack
